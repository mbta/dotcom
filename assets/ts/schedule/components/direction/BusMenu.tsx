import React, {
  ReactElement,
  Dispatch,
  KeyboardEvent as ReactKeyboardEvent
} from "react";
import {
  MenuAction,
  setRoutePatternAction,
  toggleRoutePatternMenuAction
} from "./reducer";
import { EnhancedRoutePattern } from "../__schedule";
import handleNavigation from "./menu-helpers";
import renderSvg from "../../../helpers/render-svg";
import arrowIcon from "../../../../../priv/static/icon-svg/icon-down-arrow.svg";
import checkIcon from "../../../../../priv/static/icon-svg/icon-checkmark.svg";
import { handleReactEnterKeyPress } from "../../../helpers/keyboard-events-react";

interface ExpandedBusMenuProps {
  routePatterns: EnhancedRoutePattern[];
  selectedRoutePatternId: string;
  dispatch: Dispatch<MenuAction>;
}

interface RoutePatternItem {
  routePatternIds: string[];
  routePattern: EnhancedRoutePattern;
  selected: boolean;
  duplicated: boolean;
  dispatch: Dispatch<MenuAction>;
}

interface BusMenuSelectProps {
  routePatterns: EnhancedRoutePattern[];
  selectedRoutePatternId: string;
  dispatch: Dispatch<MenuAction>;
}

const typicalityDefaultText = (typicality: number): string => {
  if (typicality === 1) return "typical route";
  if (typicality === 2) return "less common route";
  return "";
};

const RoutePatternItem = ({
  routePatternIds,
  routePattern,
  selected,
  duplicated,
  dispatch
}: RoutePatternItem): ReactElement<HTMLElement> => {
  const selectedClass = selected ? " m-schedule-direction__menu--selected" : "";

  const icon = selected ? (
    <div className="m-schedule-direction__checkmark">
      {renderSvg("c-svg__icon", checkIcon)}
    </div>
  ) : null;

  const handleClick = (): void => dispatch(setRoutePatternAction(routePattern));

  return (
    <div
      aria-current={selected ? "page" : undefined}
      id={`route-pattern_${routePattern.id}`}
      tabIndex={0}
      role="menuitem"
      className={`m-schedule-direction__menu-item${selectedClass}`}
      onClick={handleClick}
      onKeyUp={(e: ReactKeyboardEvent) => {
        handleReactEnterKeyPress(e, () => {
          handleClick();
        });
      }}
      onKeyDown={(e: ReactKeyboardEvent) => {
        handleNavigation(e, routePatternIds);
      }}
    >
      <div className="m-schedule-direction__menu-item-headsign">
        {icon}
        {routePattern.headsign}
      </div>
      <div className="m-schedule-direction__menu-item-description">
        {duplicated && `from ${routePattern.name.split(" - ")[0]}, `}
        {routePattern.time_desc ||
          typicalityDefaultText(routePattern.typicality)}
      </div>
    </div>
  );
};

const hasMoreRoutePatterns = (routePatterns: EnhancedRoutePattern[]): boolean =>
  routePatterns.some(
    (routePattern: EnhancedRoutePattern) => routePattern.typicality > 2
  );

const isDuplicateHeadsign = (
  compareRoutePattern: EnhancedRoutePattern,
  routePatterns: EnhancedRoutePattern[]
): boolean =>
  routePatterns.filter(
    (routePattern: EnhancedRoutePattern) =>
      routePattern.id !== compareRoutePattern.id &&
      routePattern.headsign === compareRoutePattern.headsign
  ).length > 0;

export const ExpandedBusMenu = ({
  routePatterns,
  selectedRoutePatternId,
  dispatch
}: ExpandedBusMenuProps): ReactElement<HTMLElement> => {
  const regularRoutePatterns = routePatterns.filter(
    (routePattern: EnhancedRoutePattern) => routePattern.typicality < 3
  );
  const uncommonRoutePatterns = routePatterns.filter(
    (routePattern: EnhancedRoutePattern) => routePattern.typicality >= 3
  );

  const routePatternIds = routePatterns.map(
    (routePattern: EnhancedRoutePattern) => routePattern.id
  );

  return (
    <div className="m-schedule-direction__menu" role="menu">
      {regularRoutePatterns.map((routePattern: EnhancedRoutePattern) => (
        <RoutePatternItem
          key={routePattern.id}
          routePatternIds={routePatternIds}
          routePattern={routePattern}
          selected={selectedRoutePatternId === routePattern.id}
          duplicated={isDuplicateHeadsign(routePattern, routePatterns)}
          dispatch={dispatch}
        />
      ))}
      {hasMoreRoutePatterns(routePatterns) && (
        <>
          <div className="m-schedule-direction__menu-item-spacer">
            Uncommon destinations
          </div>
          {uncommonRoutePatterns.map((routePattern: EnhancedRoutePattern) => (
            <RoutePatternItem
              key={routePattern.id}
              routePatternIds={routePatternIds}
              routePattern={routePattern}
              selected={selectedRoutePatternId === routePattern.id}
              duplicated={isDuplicateHeadsign(routePattern, routePatterns)}
              dispatch={dispatch}
            />
          ))}
        </>
      )}
    </div>
  );
};

const routePatternNameById = (
  routePatterns: EnhancedRoutePattern[],
  selectedRoutePatternId: string
): string =>
  routePatterns.find(
    (routePattern: EnhancedRoutePattern) =>
      routePattern.id === selectedRoutePatternId
  )!.headsign;

export const BusMenuSelect = ({
  routePatterns,
  selectedRoutePatternId,
  dispatch
}: BusMenuSelectProps): ReactElement<HTMLElement> => {
  const isMenuClickable = routePatterns.length > 1;

  const handleClick = isMenuClickable
    ? () => {
        dispatch(toggleRoutePatternMenuAction());
      }
    : /* istanbul ignore next */
      () => {};

  const linkClass = isMenuClickable
    ? " m-schedule-direction__route-pattern--clickable"
    : "";

  return (
    // eslint-disable-next-line jsx-a11y/no-static-element-interactions
    <div
      // eslint-disable-next-line jsx-a11y/no-noninteractive-tabindex
      tabIndex={isMenuClickable ? 0 : undefined}
      role={isMenuClickable ? "button" : undefined}
      className={`m-schedule-direction__route-pattern${linkClass} notranslate`}
      onClick={handleClick}
      onKeyUp={e =>
        handleReactEnterKeyPress(e, () => {
          handleClick();
        })
      }
    >
      {routePatternNameById(routePatterns, selectedRoutePatternId)}{" "}
      {isMenuClickable &&
        renderSvg(
          "c-svg__icon m-schedule-direction__route-pattern-arrow",
          arrowIcon
        )}
    </div>
  );
};
