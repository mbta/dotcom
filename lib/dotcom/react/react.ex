defmodule Dotcom.React do
  @moduledoc """
  React renderer supervisor
  """

  use Supervisor

  require Logger

  alias Phoenix.HTML

  @pool_name :react_render
  @default_pool_size 4

  @doc """
  Starts the react renderer worker pool.
  """
  @spec start_link(any) :: {:ok, pid} | {:error, any()}
  def start_link(_) do
    opts = [pool_size: @default_pool_size]
    Supervisor.start_link(__MODULE__, opts, name: __MODULE__)
  end

  @spec render(String.t(), map) :: HTML.safe()
  def render(react_component_name, args) do
    case do_render(react_component_name, args) do
      {:ok, %{"markup" => body}} ->
        body
        # Replace zero-width space HTML entities with the Unicode character
        |> String.replace("&#8203;", "â€‹")
        |> HTML.raw()

      {:error, %{"error" => %{"message" => message}}} ->
        _ =
          Logger.warning(fn -> "react_renderer component=#{react_component_name} #{message}" end)

        ""
    end
  end

  defp do_render(component, props) do
    task =
      Task.async(fn ->
        :poolboy.transaction(
          @pool_name,
          fn pid -> GenServer.call(pid, {:render, component, props}) end,
          :infinity
        )
      end)

    Task.await(task)
  end

  @doc """
  Initialize the pool supervisor
  """
  def init(opts) do
    config = Application.get_env(:dotcom, :react)

    :ok = dev_build(config[:source_path])

    pool_size = Keyword.fetch!(opts, :pool_size)

    pool_opts = [
      name: {:local, @pool_name},
      worker_module: Dotcom.React.Worker,
      size: pool_size,
      max_overflow: 0
    ]

    children = [
      :poolboy.child_spec(@pool_name, pool_opts)
    ]

    opts = [strategy: :one_for_one]
    Supervisor.init(children, opts)
  end

  def dev_build(path, cmd_fn \\ &System.cmd/3)
  def dev_build(nil, _), do: :ok

  def dev_build(path, cmd_fn) do
    {_, 0} =
      cmd_fn.(
        "npm",
        ["run", "webpack:build:react"],
        cd: path
      )

    :ok
  end
end
