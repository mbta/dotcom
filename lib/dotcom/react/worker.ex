defmodule Dotcom.React.Worker do
  @moduledoc """
  React renderer worker
  """

  use GenServer

  require Logger

  def start_link(opts \\ []) do
    GenServer.start_link(__MODULE__, opts)
  end

  def init(_opts) do
    node = System.find_executable("node") || System.find_executable("nodejs")
    config = Application.get_env(:dotcom, :react)
    port = Port.open({:spawn_executable, node}, args: [config[:build_path]])
    {:ok, %{port: port}}
  end

  def handle_call({:render, name, props}, _from, %{port: port} = state) do
    body =
      Poison.encode!(%{
        name: name,
        props: props
      })

    Port.command(port, body <> "\n")

    response =
      ""
      |> receive_response(&handle_json/1)

    if get_in(response, [:data]) do
      if get_in(response, [:data, "error"]) do
        {:reply, {:error, response.data}, state}
      else
        {:reply, {:ok, response.data}, state}
      end
    else
      {:noreply, state}
    end
  end

  defp receive_response(prev_responses, handle_fn) do
    receive do
      {_, {:data, data}} ->
        str = to_string(data)
        resp = prev_responses <> str
        handle_response(resp, handle_fn)
    end
  end

  def handle_response(resp, handle_fn) do
    {time, result} =
      :timer.tc(fn ->
        # Split and keep delimiter
        ~r/(?<=\n)/
        |> Regex.split(resp, trim: true)
        |> Enum.map(&handle_data(&1, handle_fn))
        |> Enum.find(&(!is_nil(&1)))
      end)

    _ = Logger.info("node_logging req_time milliseconds=#{time / 1_000}")
    result
  end

  def handle_data(str, handle_fn) do
    if String.ends_with?(str, "\n") do
      if String.starts_with?(str, "node_logging") do
        handle_logging(str)
      else
        handle_fn.(str)
      end
    else
      receive_response(str, handle_fn)
    end
  end

  def handle_json(msg) do
    json = Poison.decode!(msg)
    %{data: json}
  end

  def handle_logging(msg) do
    _ = Logger.info(inspect(msg))
    nil
  end

  def handle_info(msg, state) do
    _ = Logger.info(inspect(msg))
    {:noreply, state}
  end
end
