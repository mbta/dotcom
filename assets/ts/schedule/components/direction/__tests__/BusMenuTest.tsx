import React, { Dispatch } from "react";
import { render, fireEvent, screen } from "@testing-library/react";
import { EnhancedRoutePattern } from "../../__schedule";
import { BusMenuSelect, ExpandedBusMenu } from "../BusMenu";
import {
  MenuAction,
  setRoutePatternAction,
  toggleRoutePatternMenuAction
} from "../reducer";

const mockDisptach: Dispatch<MenuAction> = jest.fn();

const routePatterns: EnhancedRoutePattern[] = [
  {
    direction_id: 0,
    headsign: "Harvard via Allston",
    id: "66-6-0",
    name: "Dudley Station - Harvard Square",
    representative_trip_id: "44172015",
    representative_trip_polyline: "qwerty123@777njhgb",
    stop_ids: ["123", "456", "789"],
    route_id: "66",
    shape_id: "660140",
    time_desc: null,
    typicality: 1,
    sort_order: 5,
    canonical: false
  },
  {
    typicality: 3,
    time_desc: "School days only",
    shape_id: "660141-2",
    route_id: "66",
    representative_trip_id: "43773700_2",
    representative_trip_polyline: "lkjhg987bvcxz88!",
    stop_ids: ["123", "555", "789"],
    name: "Dudley Station - Union Square, Boston",
    id: "66-B-0",
    headsign: "Watertown Yard via Union Square Allston",
    direction_id: 0,
    sort_order: 7,
    canonical: false
  }
];
const singleRoutePattern = routePatterns.slice(0, 1);

describe("BusMenuSelect", () => {
  it("renders a menu with a single route pattern", () => {
    const { asFragment } = render(
      <BusMenuSelect
        routePatterns={singleRoutePattern}
        selectedRoutePatternId="66-6-0"
        dispatch={mockDisptach}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders a menu with multiple route patterns", () => {
    const { asFragment } = render(
      <BusMenuSelect
        routePatterns={routePatterns}
        selectedRoutePatternId="66-6-0"
        dispatch={mockDisptach}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("it does nothing when clicked if there is only one route pattern", () => {
    const { container } = render(
      <BusMenuSelect
        routePatterns={singleRoutePattern}
        selectedRoutePatternId="66-6-0"
        dispatch={mockDisptach}
      />
    );

    fireEvent.click(
      container.querySelector(".m-schedule-direction__route-pattern")!
    );
    expect(mockDisptach).not.toHaveBeenCalled();
  });

  it("it opens the route pattern menu when clicked if there are multiple route patterns", () => {
    const { container } = render(
      <BusMenuSelect
        routePatterns={routePatterns}
        selectedRoutePatternId="66-6-0"
        dispatch={mockDisptach}
      />
    );

    fireEvent.click(
      container.querySelector(".m-schedule-direction__route-pattern")!
    );
    expect(mockDisptach).toHaveBeenCalledWith(toggleRoutePatternMenuAction());

    // coverage
    fireEvent.keyUp(
      container.querySelector(".m-schedule-direction__route-pattern")!,
      { key: "Enter" }
    );
    fireEvent.keyDown(
      container.querySelector(".m-schedule-direction__route-pattern")!,
      { key: "Tab" }
    );
  });
});

describe("ExpandedBusMenu", () => {
  it("renders a menu", () => {
    const { asFragment } = render(
      <ExpandedBusMenu
        routePatterns={routePatterns}
        selectedRoutePatternId="66-6-0"
        dispatch={mockDisptach}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("selects a new route pattern when an item is clicked on", () => {
    const { container } = render(
      <ExpandedBusMenu
        routePatterns={routePatterns}
        selectedRoutePatternId="66-6-0"
        dispatch={mockDisptach}
      />
    );

    expect(
      container.querySelectorAll(".m-schedule-direction__menu-item")
    ).toHaveLength(2);

    const routePattern = routePatterns[1];
    expect(
      container.querySelectorAll(`#route-pattern_${routePattern.id}`)
    ).toHaveLength(1);

    fireEvent.click(
      container.querySelector(`#route-pattern_${routePattern.id}`)!
    );

    expect(mockDisptach).toHaveBeenCalledWith(
      setRoutePatternAction(routePattern)
    );

    fireEvent.keyUp(
      container.querySelector(`#route-pattern_${routePattern.id}`)!,
      { key: "Enter" }
    );
    fireEvent.keyDown(
      container.querySelector(`#route-pattern_${routePattern.id}`)!,
      { key: "Enter" }
    );

    const menu = screen.getByRole("menu");
    fireEvent.keyUp(menu, { key: "Enter" });
    fireEvent.keyDown(menu, { key: "Tab" });
  });
});
