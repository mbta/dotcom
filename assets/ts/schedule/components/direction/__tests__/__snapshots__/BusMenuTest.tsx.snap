// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BusMenuSelect renders a menu with a single route pattern 1`] = `
<DocumentFragment>
  <div
    class="m-schedule-direction__route-pattern notranslate"
  >
    Harvard via Allston 
  </div>
</DocumentFragment>
`;

exports[`BusMenuSelect renders a menu with multiple route patterns 1`] = `
<DocumentFragment>
  <div
    class="m-schedule-direction__route-pattern m-schedule-direction__route-pattern--clickable notranslate"
    role="button"
    tabindex="0"
  >
    Harvard via Allston 
    <span
      aria-hidden="true"
      class="notranslate c-svg__icon m-schedule-direction__route-pattern-arrow"
    >
      SVG
    </span>
  </div>
</DocumentFragment>
`;

exports[`ExpandedBusMenu renders a menu 1`] = `
<DocumentFragment>
  <div
    class="m-schedule-direction__menu"
    role="menu"
  >
    <div
      aria-current="page"
      class="m-schedule-direction__menu-item m-schedule-direction__menu--selected"
      id="route-pattern_66-6-0"
      role="menuitem"
      tabindex="0"
    >
      <div
        class="m-schedule-direction__menu-item-headsign"
      >
        <div
          class="m-schedule-direction__checkmark"
        >
          <span
            aria-hidden="true"
            class="notranslate c-svg__icon"
          >
            SVG
          </span>
        </div>
        Harvard via Allston
      </div>
      <div
        class="m-schedule-direction__menu-item-description"
      >
        typical route
      </div>
    </div>
    <div
      class="m-schedule-direction__menu-item-spacer"
    >
      Uncommon destinations
    </div>
    <div
      class="m-schedule-direction__menu-item"
      id="route-pattern_66-B-0"
      role="menuitem"
      tabindex="0"
    >
      <div
        class="m-schedule-direction__menu-item-headsign"
      >
        Watertown Yard via Union Square Allston
      </div>
      <div
        class="m-schedule-direction__menu-item-description"
      >
        School days only
      </div>
    </div>
  </div>
</DocumentFragment>
`;
